### PR TITLE
Add ValidateNullOrEmpty to -Name parameter of Get-Service

### PR DESCRIPTION
**Pull request by HemantMahawar**
Fix for #2540

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    if ( -not $IsWindows ) {
+        $PSDefaultParameterValues["it:skip"] = $true
+    }
+  }
+  # Restore the defaults
+  AfterAll {
+      $global:PSDefaultParameterValues = $originalDefaultParameterValues
+  }
+  Context 'Check null or empty value to the -Name parameter' {
+    It 'Should throw if null is passed to -Name parameter' {
```
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1]: You can use the `-Testcases` parameter to shorten this test a bit, e.g.

```powershell
$testCases =
    @{ data = $null; value = 'null' },
    @{ data = [String]::Empty, value = 'empty string' }

It 'Should throw if <value> is passed to -Name parameter' -TestCases $testCases {
    param($data)
    try { Get-Service -Name $data }
    catch [System.Management.Automation.ParameterBindingException] {$_.Exception.ErrorId | Should Be 'ParameterArgumentValidationError'}
}

It 'Should throw if <value> is passed to -Name parameter via pipeline' -TestCases $testCases {
    param($data)
    try { $data | Get-Service }
    catch [System.Management.Automation.ParameterBindingException] {$_.Exception.ErrorId | Should Be 'ParameterArgumentValidationError'}
}

```

**Comment by HemantMahawar** [comment]
@lzybkr Thanks for the great tip. Incorporated your feedback.

**Review by lzybkr — APPROVED** [review]

**Comment by vors** [comment]
@HemantMahawar it looks like GitHub cannot attribute commits to you. It's done via email on the commits. Can you add this email address to your email addresses in the GitHub settings?

**Comment by HemantMahawar** [comment]
@vors Done

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      param($data)
+      try {Get-Service -Name $data }
+      catch [System.Management.Automation.ParameterBindingException] {$_.Exception.ErrorId | Should Be 'ParameterArgumentValidationError'}
+    }
+  }
+  Context 'Check null or empty value to the -Name parameter via pipeline' {
+    It 'Should throw if <value> is passed through pipeline to -Name parameter' -TestCases $testCases {
+      param($data)
+      try {$data | Get-Service -ErrorAction Stop}
+      catch [System.Management.Automation.ParameterBindingException] {$_.Exception.ErrorId | Should Be 'ParameterArgumentValidationError'}
```
Comment (vors) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1]: Please, use [Should Throw](https://github.com/pester/Pester/wiki/Should#throw) instead.
The current code will report success, if no error is thrown (the only check happens in the catch statement).

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    }
+  }
+  # Restore the defaults
+  AfterAll {
+      $global:PSDefaultParameterValues = $originalDefaultParameterValues
+  }
+
+  $testCases = 
+    @{ data = $null; value = 'null' },
+    @{ data = [String]::Empty; value = 'empty string' }
```
Comment (vors) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1]: Where is `value` used in the test cases?

**Review by vors — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1 (hunk 3 above)
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1]: Isn't `<value>` in the description a reference to the variable?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1 (hunk 2 above)
Comment (lzybkr) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1]: Oops, I missed that - the pattern is to `throw "Expected error on previous command"` at the end of the `try` when you want to be specific about the exception thrown.

Comment (HemantMahawar) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1]: @vors @lzybkr Thanks for pointing out the holes in test (I should have caught them by running on PS 5.1 :)). Fixed

Comment (HemantMahawar) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Service.Tests.ps1]: @vors @lzybkr Thanks for pointing out the holes in test (I should have caught them by running on PS 5.1 :)). Fixed

**Comment by HemantMahawar** [comment]
@vors @lzybkr Thanks for pointing out the holes in test (I should have caught them by running on PS 5.1 :)). Fixed

**Pull request merged**